Zybo: Fix for Zynq7000 and use ps7 as submodule for Soft-CPUs

mkuhn99 opened the pull request:
(This pull-request is the second attempt of  https://github.com/litex-hub/litex-boards/pull/423)
As mentioned in the old pull-request the Zynq7000 could not be build on the Zybo-Z7-20.  To fix this I added two variants for the Zybo platform digilent_zybo_z7.py, so that the correct fpga device is selected. 
Furthermore I added a config for the ps7. As I don't have a zybo-z7-10, I only made a config for the z7-20 and the target is still using the downloaded .xci file for z7-10. 

For the target I fixed the Zynq7000 integration by introducing Memory Regions and a finalize-function like its done for the arty-z7. 

Additionally I added the "--with-ps7" option, which initializes the processing_system of the zybo board and connects it to a soft CPU via the S_GP0 AXI bus, so that all the Modules connected to this bus (UART, DDR, ...) can be accessed by the soft CPU.
Currently I only tested the DDR and added it as the main-ram, but in the future I want to use more of the Peripherals.

Issues:
-  test on zybo-z7-10
- currently all the slaves connected to the S_GP0 Bus are in the same SoC Region. It would be beneficial to have two regions for DDR and IO.
-  add software support for IO

trabucayre commented on litex_boards/platforms/digilent_zybo_z7.py:
```diff
@@ -5,7 +5,7 @@
 
 from litex.build.generic_platform import *
-from litex.build.xilinx import Xilinx7SeriesPlatform, VivadoProgrammer
+from litex.build.xilinx import Xilinx7SeriesPlatform, VivadoProgrammer, XilinxPlatform
```
Not sure it's required: Xilinx7SeriesPlatform is a XilinxPlatform's subclass

trabucayre commented on litex_boards/platforms/digilent_zybo_z7.py:
```diff
@@ -80,22 +80,45 @@
 _connectors = [
     ("pmoda", "N15 L14 K16 K14 N16 L15 J16 J14"), # XADC
-    ("pmodb", "T20 U20 V20 W20 Y18 Y19 W18 W19"),
+    ("pmodb", "V8 W8 U7 V7 Y7 Y6 V6 W6"),
```
maybe some space to align pins to others pmods?

trabucayre commented on litex_boards/platforms/digilent_zybo_z7.py:
```diff
@@ -80,22 +80,45 @@
 ]
 
+ps7_config = {
+    "z7-20" : {
+        "PCW_UIPARAM_DDR_PARTNO " : "MT41K256M16 RE-125",
+        "PCW_FPGA_FCLK0_ENABLE " : "1",
+        "PCW_UART1_BAUD_RATE " :  "115200",
+        "PCW_EN_UART1 " :  "1",
+        "PCW_UART1_PERIPHERAL_ENABLE " :  "1",
+        "PCW_UART1_UART1_IO " :  "MIO 48 .. 49",
+        "PCW_PRESET_BANK1_VOLTAGE " : "LVCMOS 1.8V",
+        "PCW_USE_M_AXI_GP0 " : "1",
+        "PCW_USE_S_AXI_GP0 " : "1",
+        "PCW_USB0_PERIPHERAL_ENABLE " : "1",
+        "PCW_USB0_USB0_IO " : "MIO 28 .. 39", 
+        "PCW_USB0_RESET_ENABLE " : "1",
+        "PCW_USB0_RESET_IO " : "MIO 46",
+        "PCW_EN_USB0 " : "1"
+    }
```
value indent?
No config for `z7-10` ? May have an exception at Plaform __init__ ?

trabucayre commented on litex_boards/platforms/digilent_zybo_z7.py:
```diff
@@ -80,22 +80,45 @@
+            "z7-10": "xc7z010-clg400-1",
+            "z7-20": "xc7z020-clg400-1"
+        }[variant]
+        XilinxPlatform.__init__(self, device, _io,  _connectors, toolchain=toolchain)
```
`s/XilinxPlatform/Xilinx7SeriesPlatform/g`

trabucayre commented on litex_boards/platforms/digilent_zybo_z7.py:
```diff
@@ -80,22 +80,45 @@
-    def __init__(self, toolchain="vivado"):
-        Xilinx7SeriesPlatform.__init__(self, "xc7z010-clg400-1", _io,  _connectors, toolchain=toolchain)
+    def __init__(self, variant="z7-20", toolchain="vivado"):
+        device = {
```
long line ?

trabucayre commented on litex_boards/targets/xilinx_zybo_z7.py:
```diff
@@ -43,55 +41,226 @@ def __init__(self, platform, sys_clk_freq, use_ps7_clk=False):
 
 class BaseSoC(SoCCore):
-    def __init__(self, sys_clk_freq=100e6, with_led_chaser=True, **kwargs):
+    def __init__(self, sys_clk_freq=int(100e6), variant="z7-10", with_ps7=False, with_led_chaser=True, **kwargs):
```
`int()` has been dropped by commit 9e7079c4c8045092aea25fba4b396bccf2ea782b
With LiteX commit 01b9ae7894eb081258113476e58c6ac21739a11c `int()` is done at soc.py level.
`with_ps7` is not clear to me

trabucayre commented on litex_boards/targets/xilinx_zybo_z7.py:
```diff
@@ -43,55 +41,226 @@ def __init__(self, platform, sys_clk_freq, use_ps7_clk=False):
+            }
         SoCCore.__init__(self, platform, sys_clk_freq, ident="LiteX SoC on Zybo Z7", **kwargs)
-
+        
```
white space

trabucayre commented on litex_boards/targets/xilinx_zybo_z7.py:
```diff
@@ -43,55 +41,226 @@ def __init__(self, platform, sys_clk_freq, use_ps7_clk=False):
-            os.makedirs("xci", exist_ok=True)
-            os.system("mv zybo_z7_ps7.txt xci/zybo_z7_ps7.xci")
-            self.cpu.set_ps7_xci("xci/zybo_z7_ps7.xci")
+            self.cpu.use_rom = True
+            
```
long line

trabucayre commented on litex_boards/targets/xilinx_zybo_z7.py:
```diff
@@ -43,55 +41,226 @@ def __init__(self, platform, sys_clk_freq, use_ps7_clk=False):
-    from litex.build.parser import LiteXArgumentParser
-    parser = LiteXArgumentParser(platform=digilent_zybo_z7.Platform, description="LiteX SoC on Zybo Z7.")
-    parser.add_target_argument("--sys-clk-freq", default=100e6, type=float, help="System clock frequency.")
+    from litex.soc.integration.soc import LiteXSoCArgumentParser
+    parser = LiteXSoCArgumentParser(description="LiteX SoC on Zybo Z7")
+    target_group = parser.add_argument_group(title="Target options")
+    target_group.add_argument("--build",        action="store_true", help="Build design.")
+    target_group.add_argument("--load",         action="store_true", help="Load bitstream.")
+    target_group.add_argument("--sys-clk-freq", default=125e6,       help="System clock frequency.")
+    target_group.add_argument("--variant",      default="z7-10",     help="Board variant (z7-10 or z7-20).")
+    target_group.add_argument("--with-ps7",     action="store_true", help="Add the PS as slave.")
+    builder_args(parser)
+    soc_core_args(parser)
```
Why revert `LiteXArgumentParser` to `LiteXArgumentParser` ?

trabucayre commented on litex_boards/targets/xilinx_zybo_z7.py:
```diff
@@ -43,55 +41,226 @@ def __init__(self, platform, sys_clk_freq, use_ps7_clk=False):
     soc = BaseSoC(
-        sys_clk_freq = args.sys_clk_freq,
-        **parser.soc_argdict
+        sys_clk_freq = int(float(args.sys_clk_freq)),
```
ditto

trabucayre commented on litex_boards/targets/xilinx_zybo_z7.py:
```diff
@@ -43,55 +41,226 @@ def __init__(self, platform, sys_clk_freq, use_ps7_clk=False):
+        sys_clk_freq = int(float(args.sys_clk_freq)),
+        variant = args.variant,
+        with_ps7 = args.with_ps7,
+        **soc_core_argdict(args)
```
ditto

trabucayre commented:
Thanks for the PR: most of my comments are nitpicking and globally looks good.
Just a last question: why `add_ps7` and `add_axi_gp_slave`: both are more or less identicals to ones in *litex/soc/cores/cpu/zynq7000/core.py*.

cklarhorst commented:
`add_axi_gp_slave` is not implemented in https://github.com/enjoy-digital/litex/blob/6b4696e3e047208530faceb8e8ddaa7aa8598549/litex/soc/cores/cpu/zynq7000/core.py#L273

The code should maybe be moved to that place. Zynq7000 currently only has HP-Slave support while that one is faster, I think only GP-Slaves allow accessing PS I/O Peripherals.

The overall idea was to support using the PS UART from soft CPUs like vexriscv.

trabucayre commented:
In my opinion: it makes more sense to complete zynq7000 support -> it's will allowing all targets based on this device to takes profits to that without having to copy/paste this code.

mkuhn99 commented:
I think i fixed all your remarks. 

> In my opinion: it makes more sense to complete zynq7000 support -> it's will allowing all targets based on this device to takes profits to that without having to copy/paste this code.

I opened a new [pull request](https://github.com/enjoy-digital/litex/pull/1520) which implements the axi slave and changed the code so that I removed the `add_ps7` and  `add_axi_gp_slave` functions. Instead I add the zynq7000 as a submodule and connect the slave axi to the main bus.

Open issues are as mentioned [in the first comment](https://github.com/litex-hub/litex-boards/pull/453#issue-1454874392) a proper mapping of the addresses from PL to PS and software support, but I think this will be a new pull request once I have worked on it.

trabucayre commented on litex_boards/targets/xilinx_zybo_z7.py:
```diff
@@ -19,79 +19,181 @@
     from litex.build.parser import LiteXArgumentParser
-    parser = LiteXArgumentParser(platform=digilent_zybo_z7.Platform, description="LiteX SoC on Zybo Z7.")
-    parser.add_target_argument("--sys-clk-freq", default=100e6, type=float, help="System clock frequency.")
+    parser = LiteXArgumentParser(description="LiteX SoC on Zybo Z7")
+    target_group = parser.add_argument_group(title="Target options")
+    target_group.add_argument("--build",        action="store_true", help="Build design.")
+    target_group.add_argument("--load",         action="store_true", help="Load bitstream.")
+    target_group.add_argument("--sys-clk-freq", default=125e6,       help="System clock frequency.")
+    target_group.add_argument("--variant",      default="z7-10",     help="Board variant (z7-10 or z7-20).")
+    target_group.add_argument("--with-ps7",     action="store_true", help="Add the PS7 as slave for soft CPUs.")
+    builder_args(parser)
+    soc_core_args(parser)
```
With `LiteXArgumentParser` **build** and **load** are automatically added -> no need to add again

By using `add_target_argument` no needs to use `target_group = parser.add_argument_group(title="Target options")`

so extras options may looks like:

```
    parser.add_target_argument("--sys-clk-freq", default=100e6, type=float, help="System clock frequency.")
    parser.add_target_argument("--variant",      default="z7-10",     help="Board variant (z7-10 or z7-20).")
    parser.add_target_argument("--with-ps7",     action="store_true", help="Add the PS7 as slave for soft CPUs.")
```
`builder_args` and `soc_core_args` are also handled by `LiteXArgumentParser` so no need to add again

trabucayre commented:
It's just a question before merging and not a request.
I have compared both flavor [board_files](https://github.com/Digilent/vivado-boards/tree/master/new/board_files): ps7 configuration is exactly the same (more or less leds and PMOD). Is the `ps7_config` is enough for your board it make sense to use it for both. Do you agree?

mkuhn99 commented:
It might work, but as I don't have a z7-10 board to test it, I would use the config only for z7-20 until someone else has tested it, so it doesn't break something.  That is why I kept the old .xci file.

trabucayre commented:
Seems more secure yes

trabucayre commented:
Thanks @mkuhn99 !
